li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaInstrumentedProducerImpl.java

```diff
@@ -0,0 +1,342 @@
+  }
+
+  @Override
+  public void beginTransaction() throws ProducerFencedException {
+    verifyOpen();
+    delegate.beginTransaction();
+  }
+
+  @Override
+  public void sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets, String consumerGroupId)
+      throws ProducerFencedException {
+    verifyOpen();
+    delegate.sendOffsetsToTransaction(offsets, consumerGroupId);
+  }
+
+  @Override
+  public void commitTransaction() throws ProducerFencedException {
+    verifyOpen();
+    delegate.commitTransaction();
+  }
+
+  @Override
+  public void abortTransaction() throws ProducerFencedException {
+    verifyOpen();
+    delegate.abortTransaction();
+  }
+
+  @Override
+  public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+    verifyOpen();
+    return delegate.send(record);
+  }
+
+  @Override
+  public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
+    verifyOpen();
+    return delegate.send(record, callback);
+  }
+
+  @Override
+  public void flush() {
+    //vanilla doesnt verify open for this method
+    Producer<K, V> delegate = this.delegate;
+    if (delegate != null) {
+      delegate.flush();
+    }
+  }
+
+  @Override
+  public List<PartitionInfo> partitionsFor(String topic) {
+    return delegate.partitionsFor(topic);
+  }
+
+  @Override
+  public Map<MetricName, ? extends Metric> metrics() {
+    //noinspection unchecked
+    return (Map<MetricName, ? extends Metric>) metricsProxy;
+  }
+
+  @Override
+  public void close() {
+    if (!proceedClosing()) {
+      return;
+    }
+    try {
+      delegate.close();
+    } finally {
+      delegate = null;
+    }
+  }
+
+  //TODO - remove after bumping up kafka-clients
+  @Override
+  public void close(long timeout, TimeUnit unit) {
+    close(Duration.ofMillis(unit.toMillis(timeout)));
+  }
+
+  //@Override
+  public void close(Duration timeout) {
+    if (!proceedClosing()) {
+      return;
+    }
+    try {
+      //TODO - fix back after bumping up kafka
+      delegate.close(timeout.toMillis(), TimeUnit.MILLISECONDS);
+    } finally {
+      delegate = null;
+    }
+  }
+
+  private boolean proceedClosing() {
+    if (isClosed()) {
+      return false;
+    }
+    delegateLock.writeLock().lock();
+    try {
+      if (isClosed()) {
+        return false;
+      }
+      closedAt = System.currentTimeMillis();
+      return true;
+    } finally {
+      delegateLock.writeLock().unlock();
+    }
+  }
+
+  private boolean isClosed() {
+    return closedAt != null;
+  }
+
+  /**
+   * throws an exception if this producer has been closed.
+   * there's also an assumption that unless closed, delegate is always != null
+   */
+  private void verifyOpen() {
+    if (isClosed()) {
+      //same exception thats thrown by a vanilla producer
+      throw new IllegalStateException("Cannot perform operation after producer has been closed");
+    }
+  }
+
+  private class MetricsProxy implements Map {
```

Comment:
Why do we need a MetricsProxy? It seems like the only usage until now, is to retrieve the metrics map, if the delegate exists.

Reply:
its possible to write code that gets the metrics map and then holds onto that map for a looong time - longer than the delegate exists. using this proxy user code that maintains the "map" for long periods of time still gets the correct sensor values even if the delegate client has been recreated since